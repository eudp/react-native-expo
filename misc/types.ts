export type FilterType = "all" | "reedemed" | "noReedemed";

export type ProductType = {
  createdAt?: string;
  product?: string;
  points?: number;
  image?: string;
  is_redemption?: boolean;
  id?: string;
};
