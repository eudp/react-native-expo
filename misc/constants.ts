import { FilterType } from "./types";

export const FILTERS = {
  ALL: "all" as FilterType,
  REEDEMED: "reedemed" as FilterType,
  NO_REEDEMED: "noReedemed" as FilterType,
};

export const API_URL =
  "https://6222994f666291106a29f999.mockapi.io/api/v1/products";

export const COLORS = {
  green: "#34C759",
  red: "#FF0000",
};
