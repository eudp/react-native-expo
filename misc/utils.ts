import { ProductType } from "./types";

export const getTotalPoints = (products: ProductType[]) =>
  products.reduce((acc, product) => {
    if (product.is_redemption) {
      return acc - product.points;
    }
    return acc + product.points;
  }, 0);
