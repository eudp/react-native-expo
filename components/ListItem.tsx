import { useMemo } from "react";
import { View, Image, StyleSheet, TouchableOpacity } from "react-native";
import moment from "moment";
import "moment/locale/es";
import { COLORS } from "../misc/constants";
import CustomText from "./CustomText";

const ListItem = ({
  image,
  product,
  createdAt,
  is_redemption,
  points,
  onPress,
}) => {
  const formattedDate = useMemo<string | undefined>(
    () => moment(createdAt).locale("es").format("DD [de] MMMM, YYYY"),
    [createdAt]
  );
  return (
    <TouchableOpacity style={styles.container} onPress={onPress}>
      <Image source={{ uri: image }} style={styles.image} />
      <View style={styles.textContainer}>
        <CustomText style={styles.name} testID="product">
          {product}
        </CustomText>
        <CustomText style={styles.date} testID="formattedDate">
          {formattedDate}
        </CustomText>
      </View>
      <View style={styles.rightContainer}>
        <View style={styles.pointsContainer}>
          <CustomText
            style={[is_redemption ? styles.red : styles.green, styles.bold]}
            testID="icon"
          >
            {is_redemption ? "-" : "+"}
          </CustomText>
          <CustomText testID="points" style={styles.bold}>
            {points}
          </CustomText>
        </View>
        <View>
          <CustomText>{">"}</CustomText>
        </View>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    borderRadius: 10,
    marginBottom: 8,
  },
  image: {
    width: 55,
    height: 55,
    marginRight: 11,
    borderRadius: 10,
  },
  textContainer: {
    flex: 1,
  },
  name: {
    lineHeight: 19,
    fontSize: 14,
    fontFamily: "Avenir-Black",
    marginBottom: 7,
  },
  date: {
    fontSize: 12,
    lineHeight: 16,
    fontWeight: "400",
  },
  rightContainer: {
    flexDirection: "row",
    alignItems: "center",
  },
  pointsContainer: {
    fontSize: 16,
    fontFamily: "Avenir-Black",
    marginRight: 17,
    lineHeight: 22,
    flexDirection: "row",
  },
  green: {
    color: COLORS.green,
  },
  red: {
    color: COLORS.red,
  },
  bold: {
    fontFamily: "Avenir-Black",
  },
});

export default ListItem;
