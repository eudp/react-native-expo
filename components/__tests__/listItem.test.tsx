import React from "react";
import { TouchableOpacity } from "react-native";
import renderer from "react-test-renderer";
import moment from "moment";
import ListItem from "../ListItem";
import { COLORS } from "../../misc/constants";

describe("ListItem", () => {
  const onPressMock = jest.fn();
  const props = {
    image: "https://example.com/image.jpg",
    product: "Test Product",
    createdAt: "2022-04-09T12:34:56.000Z",
    is_redemption: false,
    points: 100,
    onPress: onPressMock,
  };

  it("renders correctly", () => {
    const tree = renderer.create(<ListItem {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("displays the correct product name", () => {
    const component = renderer.create(<ListItem {...props} />);
    const nameText = component.root.findByProps({ testID: "product" }).props
      .children;
    expect(nameText).toBe(props.product);
  });

  it("displays the correct formatted date", () => {
    const component = renderer.create(<ListItem {...props} />);
    const formattedDateText = component.root.findByProps({
      testID: "formattedDate",
    }).props.children;
    const formattedDate = moment(props.createdAt)
      .locale("es")
      .format("DD [de] MMMM, YYYY");
    expect(formattedDateText).toBe(formattedDate);
  });

  it("displays the correct points and icon color", () => {
    const component = renderer.create(<ListItem {...props} />);
    const pointsText = component.root.findByProps({ testID: "points" }).props
      .children;
    const {
      children: icon,
      style: [{ color: iconColor }],
    } = component.root.findByProps({ testID: "icon" }).props;

    let expectedPointsText = props.points;
    let expectedIcon = "+";
    let expectedIconColor = COLORS.green;

    if (props.is_redemption) {
      expectedPointsText = props.points;
      expectedIcon = "-";
      expectedIconColor = COLORS.red;
    }
    expect(pointsText).toBe(expectedPointsText);
    expect(iconColor).toBe(expectedIconColor);
    expect(icon).toBe(expectedIcon);
  });

  it("calls onPress when pressed", () => {
    const component = renderer.create(<ListItem {...props} />);
    const listItem = component.root.findByType(TouchableOpacity);
    listItem.props.onPress();
    expect(onPressMock).toHaveBeenCalled();
  });
});
