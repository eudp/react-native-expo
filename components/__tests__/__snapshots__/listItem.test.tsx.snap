// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderRadius": 10,
      "flexDirection": "row",
      "marginBottom": 8,
      "opacity": 1,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://example.com/image.jpg",
      }
    }
    style={
      {
        "borderRadius": 10,
        "height": 55,
        "marginRight": 11,
        "width": 55,
      }
    }
  />
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Avenir-Regular",
          },
          {
            "fontFamily": "Avenir-Black",
            "fontSize": 14,
            "lineHeight": 19,
            "marginBottom": 7,
          },
        ]
      }
      testID="product"
    >
      Test Product
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "Avenir-Regular",
          },
          {
            "fontSize": 12,
            "fontWeight": "400",
            "lineHeight": 16,
          },
        ]
      }
      testID="formattedDate"
    >
      09 de abril, 2022
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "fontFamily": "Avenir-Black",
          "fontSize": 16,
          "lineHeight": 22,
          "marginRight": 17,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Avenir-Regular",
            },
            [
              {
                "color": "#34C759",
              },
              {
                "fontFamily": "Avenir-Black",
              },
            ],
          ]
        }
        testID="icon"
      >
        +
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "Avenir-Regular",
            },
            {
              "fontFamily": "Avenir-Black",
            },
          ]
        }
        testID="points"
      >
        100
      </Text>
    </View>
    <View>
      <Text
        style={
          [
            {
              "fontFamily": "Avenir-Regular",
            },
            undefined,
          ]
        }
      >
        &gt;
      </Text>
    </View>
  </View>
</View>
`;
